verification/io/io-spec.gobra

```diff
@@ -92,79 +92,92 @@ pure func dp3s_iospec_bio3s_enter_guard(s IO_dp3s_state_local, t Place, v IO_val
 				v.IO_Internal_val1_4)
 }
 
-pred dp3s_iospec_bio3s_enter(s IO_dp3s_state_local, t Place) {
+pred (dp DataPlaneSpec) dp3s_iospec_bio3s_enter(s IO_dp3s_state_local, t Place) {
 	// TODO: we may need more triggering terms here
-	forall v IO_val :: { dp3s_iospec_bio3s_enter_guard(s, t, v) } (v.isIO_Internal_val1 ==>
-			(dp3s_iospec_bio3s_enter_guard(s, t, v) ==>
+	forall v IO_val :: { dp.dp3s_iospec_bio3s_enter_guard(s, t, v) } (
+		match v {
+			case IO_Internal_val1{_, _, ?newpkt, ?nextif}: 
+				(dp.dp3s_iospec_bio3s_enter_guard(s, t, v) ==>
 				(CBio_IN_bio3s_enter(t, v) &&
-				dp3s_iospec_ordered(
-					dp3s_add_obuf(s, v.IO_Internal_val1_4, v.IO_Internal_val1_3),
-					CBio_IN_bio3s_enter_T(s, t, v)))))
+				dp.dp3s_iospec_ordered(
+					dp.dp3s_add_obuf(s, nextif, newpkt),
+					CBio_IN_bio3s_enter_T(t, v))))
+			default: true
+		})
 }
 
 pred CBio_IN_bio3s_xover_up2down(t Place, v IO_val)
 
 ghost
 requires CBio_IN_bio3s_xover_up2down(t, v)
-pure func dp3s_iospec_bio3s_xover_up2down_T(s IO_dp3s_state_local, t Place, v IO_val) Place
+pure func dp3s_iospec_bio3s_xover_up2down_T(t Place, v IO_val) Place
 
 // This corresponds to the condition of the if statement in the io-spec case for xover_up2down
 ghost
 decreases
-pure func dp3s_iospec_bio3s_xover_up2down_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
-	return let currseg := v.IO_Internal_val1_1.CurrSeg in
-	v.IO_Internal_val1_1.LeftSeg == none[IO_seg2] ? false :
-	let nextseg := get(v.IO_Internal_val1_1.LeftSeg) in
-	(nextseg.ConsDir &&
-	len(nextseg.Future) > 0 &&
-	len(currseg.Future) > 0 &&
-	let hf1, hf2 := currseg.Future[0], nextseg.Future[0] in
-	let traversedseg := dp3s_iospec_bio3s_create_guard_traversedseg(currseg) in
-	(xover_up2down2_link_type(asid(), hf1, hf2) &&
-		dp3s_xover_common(
-			s,
-			v.IO_Internal_val1_1,
-			currseg,
-			nextseg,
-			traversedseg,
-			IO_pkt2(IO_Packet2{nextseg, v.IO_Internal_val1_1.MidSeg, v.IO_Internal_val1_1.RightSeg, some(traversedseg)}),
-			hf1,
-			hf2,
-			v.IO_Internal_val1_2,
-			v.IO_Internal_val1_3,
-			v.IO_Internal_val1_4,)))
+pure func (dp DataPlaneSpec) dp3s_iospec_bio3s_xover_up2down_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
+	return let currseg := v.IO_Internal_val1_1.CurrSeg in 
```

Comment:
Is v guaranteed to be a `IO_Internal_val1`? If not, should we match on v as well? I guess that the default case (when it is not IO_Internal_val1) would be false?

Reply:
Yes v is guaranteed to be a `IO_Internal_val1`.

Reply:
Can we add a precondition making sure that this is the case?

Reply:
Sure.